cpp/src/datetime/datetime_ops.cu

```diff
@@ -290,6 +363,70 @@ std::unique_ptr<column> add_calendrical_months(column_view const& timestamp_colu
   return output;
 }
 
+std::unique_ptr<column> ceil_general(column_view const& column,
+                                     datetime_component Component,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr)
+{
+  switch (Component) {
+    case datetime_component::DAY:
+      return cudf::type_dispatcher(column.type(),
+                                   dispatch_ceil<detail::ceil_timestamp<datetime_component::DAY>>{},
+                                   column,
+                                   stream,
+                                   mr);
+    case datetime_component::HOUR:
+      return cudf::type_dispatcher(
+        column.type(),
+        dispatch_ceil<detail::ceil_timestamp<datetime_component::HOUR>>{},
+        column,
+        stream,
+        mr);
+    case datetime_component::MINUTE:
+      return cudf::type_dispatcher(
+        column.type(),
+        dispatch_ceil<detail::ceil_timestamp<datetime_component::MINUTE>>{},
+        column,
+        stream,
+        mr);
+    case datetime_component::SECOND:
+      return cudf::type_dispatcher(
+        column.type(),
+        dispatch_ceil<detail::ceil_timestamp<datetime_component::SECOND>>{},
+        column,
+        stream,
+        mr);
+    case datetime_component::MILLISECOND:
+      return cudf::type_dispatcher(
+        column.type(),
+        dispatch_ceil<detail::ceil_timestamp<datetime_component::MILLISECOND>>{},
+        column,
+        stream,
+        mr);
+    case datetime_component::MICROSECOND:
+      return cudf::type_dispatcher(
+        column.type(),
+        dispatch_ceil<detail::ceil_timestamp<datetime_component::MICROSECOND>>{},
+        column,
+        stream,
+        mr);
+    case datetime_component::NANOSECOND:
+      return cudf::type_dispatcher(
+        column.type(),
+        dispatch_ceil<detail::ceil_timestamp<datetime_component::NANOSECOND>>{},
+        column,
+        stream,
+        mr);
+    default:
+      return cudf::type_dispatcher(
+        column.type(),
+        dispatch_ceil<detail::ceil_timestamp<datetime_component::SECOND>>{},
```

Comment:
Here the default case should never reach. If yes, there should be an error thus we should call `CUDF_FAIL`.